examples/medplum-scheduling-demo/src/components/AppointmentActions.tsx

```diff
@@ -0,0 +1,135 @@
+import { Button, Stack, Title } from '@mantine/core';
+import { IconCancel, IconCircleCheck, IconCircleOff } from '@tabler/icons-react';
+import { Appointment, Encounter, Patient, Practitioner } from '@medplum/fhirtypes';
+import { Loading, useMedplum, useMedplumProfile } from '@medplum/react';
+import { showNotification } from '@mantine/notifications';
+import { useNavigate } from 'react-router-dom';
+import { createReference, normalizeErrorString } from '@medplum/core';
+import { RescheduleAppointment } from './RescheduleAppointment';
+import { useDisclosure } from '@mantine/hooks';
+import { useEffect, useState } from 'react';
+
+interface AppointmentActionsProps {
+  appointment: Appointment;
+  patient: Patient;
+}
+
+export function AppointmentActions(props: AppointmentActionsProps): JSX.Element {
+  const { appointment, patient } = props;
+  const medplum = useMedplum();
+  const profile = useMedplumProfile() as Practitioner;
+  const navigate = useNavigate();
+  const [rescheduleOpened, rescheduleHandlers] = useDisclosure(false);
+  const [encounter, setEncounter] = useState<Encounter | undefined | boolean>(false); // `false` means it was not loaded yet
+
+  async function refreshEncounter(): Promise<void> {
+    try {
+      const result = await medplum.searchOne('Encounter', { appointment: `Appointment/${appointment.id}` });
+      setEncounter(result);
+    } catch (err) {
+      console.error(err);
+    }
+  }
+
+  useEffect(() => {
+    refreshEncounter().catch(console.error);
+  }, []);
+
+  if (!appointment) {
+    return <Loading />;
+  }
+
+  // Handler for completing or cancelling the appointment
+  async function handleChangeStatus(newStatus: Appointment['status']): Promise<void> {
+    try {
+      await medplum.updateResource({
+        ...appointment,
+        status: newStatus,
+      });
+
+      navigate(`/Appointment/${appointment.id}/details`);
+      showNotification({
+        icon: <IconCircleCheck />,
+        title: 'Success',
+        message: 'Appointment status updated',
+      });
+    } catch (err) {
+      showNotification({
+        icon: <IconCircleOff />,
+        title: 'Error',
+        message: normalizeErrorString(err),
+      });
+    }
+  }
+
+  async function createEncounter(): Promise<void> {
+    try {
+      await medplum.createResource({
+        resourceType: 'Encounter',
+        status: 'finished',
+        subject: createReference(patient),
+        appointment: [createReference(appointment)],
+        class: {
+          system: 'http://terminology.hl7.org/CodeSystem/v3-ActCode',
+          code: 'AMB',
+          display: 'ambulatory',
+        },
+        serviceType: appointment.serviceType?.[0],
+        period: {
+          start: appointment.start,
+          end: appointment.end,
+        },
+        participant: [
```

Comment:
nit: can you also populate `Encounter.length`. While this is easily calculable from start/end, it's a key field needed for billing and is also searchable.